### PR TITLE
feat: Fix retain cycles in Amplitude so instance will not leak memory

### PR DESCRIPTION
**Pull request by crleona**
### Summary

Fixes retain cycles internal to the Amplitude object, allowing it to correctly deallocate. (see https://github.com/amplitude/Amplitude-Swift/issues/44).

We also add a test ensuring that Amplitude will correctly deallocate to prevent this from happening in the future.

### Checklist

* [X] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-SDK-Template/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:  No

### CODE REVIEW

**Review by izaaz — APPROVED** [review]

File: Sources/Amplitude/Plugins/AnalyticsConnectorPlugin.swift [hunk 1]
```diff
@@ -8,8 +8,10 @@ class AnalyticsConnectorPlugin: BeforePlugin {
     override func setup(amplitude: Amplitude) {
         super.setup(amplitude: amplitude)
         connector = AnalyticsConnector.getInstance(amplitude.configuration.instanceName)
-        connector!.eventBridge.setEventReceiver { event in
-            amplitude.track(event: BaseEvent(
+        let logger = amplitude.logger
+        connector!.eventBridge.setEventReceiver { [weak amplitude, logger] event in
+            logger?.error(message: "Amplitude instance has been deallocated, please maintain a strong reference to track events from Experiment")
```
Comment (crleona) [inline Sources/Amplitude/Plugins/AnalyticsConnectorPlugin.swift:13]: @izaaz added an error message if we start to receive events from Experiment but the amplitude instance was deallocated. This does maintain a strong reference to logger, but this should be better than the entire amplitude instance.

**Review by izaaz — APPROVED** [review]

File: Sources/Amplitude/Plugins/NetworkConnectivityCheckerPlugin.swift [hunk 2]
```diff
@@ -56,12 +56,15 @@ open class NetworkConnectivityCheckerPlugin: BeforePlugin {
         pathCreation.start(queue: amplitude.trackingQueue)
         pathUpdateCancellable = pathCreation.networkPathPublisher?
             .sink(receiveValue: { [weak self] networkPath in
+                guard let amplitude = self?.amplitude else {
+                    return
```
Comment (justin-fiedler) [inline Sources/Amplitude/Plugins/NetworkConnectivityCheckerPlugin.swift:60]: Maybe a debug warning here?

**Review by justin-fiedler — APPROVED** [review]
Thanks @crleona

File: Sources/Amplitude/Plugins/NetworkConnectivityCheckerPlugin.swift (hunk 2 above)
Comment (crleona) [reply Sources/Amplitude/Plugins/NetworkConnectivityCheckerPlugin.swift:62]: added!

**Pull request merged**